File: app/src/main/java/com/oliver/weatherapp/screens/home/CitiesDiffCallback.java
package com.oliver.weatherapp.screens.home;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.util.DiffUtil;

import com.oliver.weatherapp.data.local.model.CityEntry;

import java.util.List;

public class CitiesDiffCallback extends DiffUtil.Callback {

    private List<CityEntry> oldData;
    private List<CityEntry> newData;

    public CitiesDiffCallback(@NonNull List<CityEntry> oldData, @NonNull List<CityEntry> newData) {
        this.oldData = oldData;
        this.newData = newData;
    }

    @Override
    public int getOldListSize() {
        return oldData.size();
    }

    @Override
    public int getNewListSize() {
        return newData.size();
    }

    @Override
    public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
        return oldData.get(oldItemPosition).id == newData.get(newItemPosition).id;
    }

    @Override
    public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
        return oldData.get(oldItemPosition).equals(newData.get(newItemPosition));
    }

    @Nullable
    @Override
    public Object getChangePayload(int oldItemPosition, int newItemPosition) {
        //you can return particular field for changed item.
        return super.getChangePayload(oldItemPosition, newItemPosition);
    }
}

File: app/src/main/java/com/oliver/weatherapp/screens/home/CitiesDiffCallback.kt
package com.oliver.weatherapp.screens.home

import android.support.v7.util.DiffUtil

import com.oliver.weatherapp.data.local.model.CityEntry

class CitiesDiffCallback(
        private val oldData: List<CityEntry>,
        private val newData: List<CityEntry>
) : DiffUtil.Callback() {

    override fun getOldListSize(): Int {
        return oldData.size
    }

    override fun getNewListSize(): Int {
        return newData.size
    }

    override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        return oldData[oldItemPosition].id == newData[newItemPosition].id
    }

    override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        return oldData[oldItemPosition] == newData[newItemPosition]
    }
}